src/secret_sharing/mod.rs

```diff
@@ -51,44 +58,126 @@ pub trait Block: Sized + Copy + Debug {
     type Size: ArrayLength;
 }
 
-///allows basic secret sharing operations
-pub trait WeakSharedValue:
+/// Trait for types that are input to our additive secret sharing scheme.
+///
+/// Additive secret sharing requires an addition operation. In cases where arithmetic secret sharing
+/// (capable of supporting addition and multiplication) is desired, the `Field` trait extends
+/// `SharedValue` to require multiplication.
+pub trait SharedValue:
     Clone + Copy + PartialEq + Debug + Send + Sync + Sized + Additive + Serializable + 'static
 {
     type Storage: Block;
+    type Array<const N: usize>: SharedValueArray<Self>;
 
     const BITS: u32;
 
     const ZERO: Self;
 }
 
-///allows advanced secret sharing operations, requires multiplication
-pub trait SharedValue:
-    Clone + Copy + PartialEq + Debug + Send + Sync + Sized + Arithmetic + Serializable + 'static
-{
-    type Storage: Block;
+// The purpose of this trait is to avoid placing a `Message` trait bound on `SharedValueArray`, or
+// similar. Doing so would require either (1) a generic impl of `Serializable` for any `N`, which
+// is hard to write, or (2) additional trait bounds of something like `F::Array<1>: Message`
+// throughout many of our protocols.
+//
+// Writing `impl<F: Field> Vectorized<1> for F` means that the compiler will always see that it
+// is available anywhere an `F: Field` trait bound is effective.
 
-    const BITS: u32;
+pub trait Vectorized<const N: usize>: SharedValue {
+    type Message: Message + Clone;
```

Comment:
curious what drives the requirement for it to be `Clone`?

Reply:
Probably I had a `Clone` somewhere in one of my working implementations of vectorized multiply. I'll try to get rid of this.

Reply:
A couple things that make it hard to remove `Clone`:
* Currently the `send` function takes its argument by value, not by reference.
* We `clone` one of the arguments when what we really want is a ref-times-ref multiply.

Both of these are fixable, but not sure if we want to combine it with these changes.

Reply:
would it make it easier for now for `send` to take `B: Borrow<Message>`?